From bb79efac218559a3769eb535394fa955ce733c56 Mon Sep 17 00:00:00 2001
From: Michael Bayer <mike_mp@zzzcomputing.com>
Date: Thu, 11 Jun 2015 08:11:44 -0700
Subject: [PATCH] Add an index to virtual_interfaces.uuid

Performance issues have been reported such
that the query performed by virtual_interface_get_by_uuid
becomes prohibitively slow after a large (~250K) number of instances
have been created, and an attempt is made to launch instances
in batch.  An index added to uuid shows a dramatic improvement
to this query in EXPLAIN testing.

(cherry picked from commit 7aa4b75da9722496c4e166b44cc04973fd441f52)

Renumbered migration from 295 to 281 for downstream backport.

Upstream-Liberty: https://review.openstack.org/#/c/154243/
Closes-bug: #1445675
Resolves: rhbz1085989

Change-Id: I912e5a01ac63f2a7ab5533e9b6ff8738af819d67
Reviewed-on: https://code.engineering.redhat.com/gerrit/52654
Reviewed-by: Dan Smith <dansmith@redhat.com>
Reviewed-by: Matthew Booth <mbooth@redhat.com>
Tested-by: Matthew Booth <mbooth@redhat.com>
---
 .../281_add_virtual_interfaces_uuid_index.py       | 54 ++++++++++++++++++++++
 nova/db/sqlalchemy/models.py                       |  1 +
 nova/tests/unit/db/test_migrations.py              | 12 +++++
 3 files changed, 67 insertions(+)
 create mode 100644 nova/db/sqlalchemy/migrate_repo/versions/281_add_virtual_interfaces_uuid_index.py

diff --git a/nova/db/sqlalchemy/migrate_repo/versions/281_add_virtual_interfaces_uuid_index.py b/nova/db/sqlalchemy/migrate_repo/versions/281_add_virtual_interfaces_uuid_index.py
new file mode 100644
index 0000000..8daf92f
--- /dev/null
+++ b/nova/db/sqlalchemy/migrate_repo/versions/281_add_virtual_interfaces_uuid_index.py
@@ -0,0 +1,54 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+
+from oslo_log import log as logging
+from sqlalchemy import MetaData, Table, Index
+
+from nova.i18n import _LI
+
+LOG = logging.getLogger(__name__)
+
+INDEX_COLUMNS = ['uuid']
+INDEX_NAME = 'virtual_interfaces_uuid_idx'
+
+
+def _get_table_index(migrate_engine):
+    meta = MetaData()
+    meta.bind = migrate_engine
+    table = Table('virtual_interfaces', meta, autoload=True)
+    for idx in table.indexes:
+        if idx.columns.keys() == INDEX_COLUMNS:
+            break
+    else:
+        idx = None
+    return meta, table, idx
+
+
+def upgrade(migrate_engine):
+    meta, table, index = _get_table_index(migrate_engine)
+    if index:
+        LOG.info(_LI('Skipped adding %s because an equivalent index'
+                     ' already exists.'), INDEX_NAME)
+        return
+    columns = [getattr(table.c, col_name) for col_name in INDEX_COLUMNS]
+    index = Index(INDEX_NAME, *columns)
+    index.create(migrate_engine)
+
+
+def downgrade(migrate_engine):
+    meta, table, index = _get_table_index(migrate_engine)
+    if not index:
+        LOG.info(_LI('Skipped removing %s because no such index exists'),
+                     INDEX_NAME)
+        return
+    index.drop(migrate_engine)
diff --git a/nova/db/sqlalchemy/models.py b/nova/db/sqlalchemy/models.py
index f2700ff..da50931 100644
--- a/nova/db/sqlalchemy/models.py
+++ b/nova/db/sqlalchemy/models.py
@@ -860,6 +860,7 @@ class VirtualInterface(BASE, NovaBase):
                         name="uniq_virtual_interfaces0address0deleted"),
         Index('virtual_interfaces_network_id_idx', 'network_id'),
         Index('virtual_interfaces_instance_uuid_fkey', 'instance_uuid'),
+        Index('virtual_interfaces_uuid_idx', 'uuid'),
     )
     id = Column(Integer, primary_key=True, nullable=False)
     address = Column(String(255))
diff --git a/nova/tests/unit/db/test_migrations.py b/nova/tests/unit/db/test_migrations.py
index 0cf4af8..b7894c1 100644
--- a/nova/tests/unit/db/test_migrations.py
+++ b/nova/tests/unit/db/test_migrations.py
@@ -879,6 +879,18 @@ class NovaMigrationsCheckers(test_migrations.ModelsMigrationsSync,
         key_pairs = oslodbutils.get_table(engine, 'key_pairs')
         self.assertTrue(key_pairs.c.name.nullable)
 
+    def _pre_upgrade_281(self, engine):
+        self.assertIndexNotExists(engine, 'virtual_interfaces',
+                                  'virtual_interfaces_uuid_idx')
+
+    def _check_281(self, engine, data):
+        self.assertIndexMembers(engine, 'virtual_interfaces',
+                                'virtual_interfaces_uuid_idx', ['uuid'])
+
+    def _post_downgrade_281(self, engine):
+        self.assertIndexNotExists(engine, 'virtual_interfaces',
+                                  'virtual_interfaces_uuid_idx')
+
 
 class TestNovaMigrationsSQLite(NovaMigrationsCheckers,
                                test_base.DbTestCase,
